library/src/hcc_detail/hipsolver.cpp

```diff
@@ -697,6 +714,219 @@ catch(...)
     return exception2hip_status();
 }
 
+/******************** GESVD ********************/
+hipsolverStatus_t hipsolverSgesvd_bufferSize(hipsolverHandle_t handle, int m, int n, int* lwork)
+try
+{
+    *lwork = 0;
+    return HIPSOLVER_STATUS_NOT_SUPPORTED;
```

Comment:
Why is this not supported?
Users should be able to pass a pointer to work of size returned by this function (ie. rocsolver_gesvd_getMemorySize) with both backends

Reply:
I left this as unsupported for now because the only two arguments to gesvd's bufferSize function are `m` and `n`. I wasn't sure how to get the maximum needed memory size without querying rocsolver_gesvd using every possible combination of `left_svect` and `right_svect`.

Reply:
oh, I see!

My suggestion is to add `left_svect` and `right_svect` to the bufferSize APIs, and ignore them on the CUDA side. The call won't be exactly as in cuSOLVER, but it will work (better to have extra parameters than missing parameters).
I think this is better than having the same user code falling on the HIP side because the bufferSize API will return not_supported.

Alternatively, returning lwork=0 and a success status should be fine... 
If lwork is zero, and you do `rocblas_set_workspace` on the implementation of the gesvd function, the memory model will be switched to automatic anyways.

What do you think?

Reply:
I would rather not deviate from the cuSOLVER API by adding new arguments, as my understanding is that the point of hipBLAS and hipSOLVER is that the user should just be able to hipify their cuda code and have it run on our hardware without issue. The easier it is to turn cuSOLVER code into hipSOLVER code, the better.

I could change the returned status to be a success, sure. My only concern is that will sort of hide the fact that we always use the internal workspace with gesvd and never the user-provided workspace. On that note, I'd still prefer not to use `rocblas_set_workspace` in gesvd. If a user creates a workspace of size `max(getrf_bufferSize, gesvd_bufferSize)`, for example, it might not be large enough since `gesvd_bufferSize` is not giving the correct size. If the user passes that workspace to gesvd with `lwork != 0`, and we call `rocblas_set_workspace` on it, that will cause problems.

Reply:
No, the main purpose of hipBLAS and hipSOLVER is not hipify code (there are other tools like hipify-clang for that). hipBLAS/SOLVER are marshalling libraries; they should provide an interface that does not require the user code to change regardless of what backend (cuda or hip) is used. Ideally, to facilitate the transition of cuda libraries' users to rocm libraries, we want this interface to be as similar to cuBLAS/SOLVER as possible, but it is not a requirement.  

The requirement is that the same code, let's say:

int lwork;
T *work;
hipsolverSgesvd_bufferSize(...,*lwork);
hipMalloc(..., work, work);
hipsolverSgesvd(..., work, lwork);

works OK with both backends. This won't be the case if the user checks returned status of the buffeSize API, for example; or, as you mentioned if we return lwork=0 but the user then does lwork=max(...) with the buffer size of another function, etc (there are many ways this could fail)...

IMO, the safest bet is to have the bufferSize API receiving the svects as parameters and letting both backends do their thing.

Reply:
Interesting. If we're going to be allowing changes from the cuSOLVER API, there are other changes I would like to make, such as adding workspace arguments to getrs and potrfBatched, as well as adding an lwork argument to getrf. I'll have to update the documentation about the special considerations.

Reply:
Adding lwork to getrf is not a bad idea!... This way it would be cleaner on the hip side; no need to do the memory query on the function implementation. On the cuda side, lwork could be ignored.

On the other hand, I don't see why adding workspace to getrs or potrf calls, neither rocSOLVER nor cuSOLVER really need it.  rocSOLVER will manage workspace automatically, and cuSOLVER will do whatever it does.

Reply:
I want to add workspaces to getrs and potrfBatched so that we can avoid switching between user-provided and internal workspaces when the user wants to provide the workspace. As noted in the documentation, this switching can impact performance.